### PR TITLE
Support CARGO_BIN_EXE env vars

### PR DESCRIPTION
**Pull request by illicitonion**
Since 1.43 cargo will set `CARGO_BIN_EXE`-prefixed env vars for each binary your crate depends on, so you can access them in integration tests.

Also, output binary names with the same name as the target, rather than replacing `-`s with `_`s, to better emulate how Cargo works.

### CODE REVIEW

File: rust/private/rustc.bzl [hunk 1]
```diff
@@ -320,17 +320,34 @@ def rustc_compile_action(
...
+    maybe_rename = ""
+    if crate_info.type == "bin":
+        generated_file = crate_info.name
+        if toolchain.target_arch == "wasm32":
+            generated_file = generated_file + ".wasm"
+        src = "/".join([output_dir, generated_file])
+        dst = crate_info.output.path
+        if src != dst:
+            maybe_rename = " && /bin/mv {src} {dst}".format(src=src, dst=dst)
+    command = '{manifest_dir_env}{out_dir_env}{rustc} "$@" --remap-path-prefix="$(pwd)"=__bazel_redacted_pwd{maybe_rename}'.format(
```
Comment (damienmg) [inline rust/private/rustc.bzl:332]: This line is long, maybe use a triple quoted string?

File: rust/private/rustc.bzl [hunk 2]
```diff
@@ -320,17 +320,34 @@ def rustc_compile_action(
...
-        toolchain.rustc.path,
+    maybe_rename = ""
+    if crate_info.type == "bin":
+        generated_file = crate_info.name
+        if toolchain.target_arch == "wasm32":
+            generated_file = generated_file + ".wasm"
+        src = "/".join([output_dir, generated_file])
+        dst = crate_info.output.path
+        if src != dst:
+            maybe_rename = " && /bin/mv {src} {dst}".format(src=src, dst=dst)
```
Comment (damienmg) [inline rust/private/rustc.bzl:331]: Why do you need that maybe_rename? I cannot make sense of the logic.

Comment (illicitonion) [reply rust/private/rustc.bzl:331]: If a target name is `foo-bar`, both Cargo and rules_rust will translate this into a crate name of `foo_bar`, and so `rustc` will create an output file (in an intermediate directory) named `foo_bar`.

However, Cargo, when putting the output in your `target/${config}` directory will use the original target name of `foo-bar`.

The env var Cargo sets uses uses the original target name, with `-` not `_`.

If we'd rather not do this, we could instead just set the env var to use `foo-bar` and point at a file named `foo_bar`, but this seemed more compatible with what Cargo does.

If we want to keep doing this, I can gladly add more comments to explain :)

Comment (damienmg) [reply rust/private/rustc.bzl:331]: I am fine with this behavior but I'd gladly take some comment here :)

File: rust/private/rustc.bzl (hunk 1 above)
Comment (illicitonion) [reply rust/private/rustc.bzl]: Reverted away from the the fstring-like templates

File: rust/private/rustc.bzl (hunk 2 above)
Comment (illicitonion) [reply rust/private/rustc.bzl:338]: Added comment

**Comment by damienmg** [comment]
Thanks!

**Pull request merged**